pymongo/client_session.py

```diff
@@ -368,6 +377,97 @@ def _inherit_option(self, name, val):
+              orders.insert_one({"sku": "abc123", "qty": 100}, session=session)
+              inventory.update_one({"sku": "abc123", "qty": {"$gte": 100}},
+                                   {"$inc": {"qty": -100}}, session=session)
+
+          with client.start_session() as session:
+              session.with_transaction(callback)
+
+        In an ideal scenario, `with_transaction` will start a transaction,
+        execute the callback once, and commit the transaction. In the event of
+        error, the commit or entire transaction may need to be retried and
+        thus the callback may be invoked multiple times. Even though the
+        callback may be invoked multiple times, `with_transaction` ensures
+        that the transaction will be committed at most once on the server.
+
+        In order to safeguard applications from infinite retry loops,
+        `with_transaction` will cease retrying the callback and
+        `commit_transaction` if it has exceeded a fixed timeout period of 120
+        seconds. This limit is non-configurable and is intentionally twice the
+        value of MongoDB's default for the  `transactionLifetimeLimitSeconds`
+        parameter (60 seconds). Applications that desire longer retry periods
+        may call `with_transaction` additional times as needed. Applications
+        that desire shorter retry periods should not use this method.
+
+        :Parameters:
+          - `callback`: The callback to run inside a transaction which must
+            accept a single argument, this session. Under certain error
+            conditions the callback may be run multiple times.
+          - `read_concern` (optional): The
+            :class:`~pymongo.read_concern.ReadConcern` to use for this
+            transaction.
+          - `write_concern` (optional): The
+            :class:`~pymongo.write_concern.WriteConcern` to use for this
+            transaction.
+          - `read_preference` (optional): The read preference to use for this
+            transaction. If ``None`` (the default) the :attr:`read_preference`
+            of this :class:`Database` is used. See
+            :mod:`~pymongo.read_preferences` for options.
+
+        :Returns:
+          The return value of the callback.
+
+        .. versionadded:: 3.9
+        """
+        start_time = monotonic.time()
+        while True:
+            self.start_transaction(
+                read_concern, write_concern, read_preference)
+            try:
+                ret = callback(self)
+            except Exception as exc:
+                if self._in_transaction:
+                    self.abort_transaction()
+                if (isinstance(exc, PyMongoError) and
+                        exc.has_error_label("TransientTransactionError") and
+                        _within_time_limit(start_time)):
+                    # Retry the entire transaction.
+                    continue
+                raise
+
+            if self._transaction.state in (
```

Comment:
Should we mention in the docstring then that there is no restriction on the callback explicitly committing the transaction? Currently, the docstring gives the impression that the callback should comprise only of the operations and everything transaction-related is handled by the helper.

Reply:
Ok, having read the [relevant section of the spec](https://github.com/mongodb/specifications/blob/master/source/transactions-convenient-api/transactions-convenient-api.rst#it-is-assumed-that-the-callback-will-not-start-a-new-transaction-on-the-clientsession) it seems like there IS a limitation on the callback but it doesn't have to do with ending a transaction, it has to do with starting it. I think we should very clearly state that users should not start new transactions within `callback` in the `with_transaction` docstring.

Reply:
Thanks for pointing this out! I updated the docstring.